pinot-controller/src/main/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimator.java

```diff
@@ -0,0 +1,589 @@
+   */
+  public MemoryEstimator(TableConfig tableConfig, Schema schema, SchemaWithMetaData schemaWithMetadata,
+      int numberOfRows, double ingestionRatePerPartition, long maxUsableHostMemory, int tableRetentionHours,
+      File workingDir) {
+    this(tableConfig, generateCompletedSegment(schemaWithMetadata, schema, tableConfig, numberOfRows, workingDir),
+        ingestionRatePerPartition, maxUsableHostMemory, tableRetentionHours, workingDir);
+  }
+
+  /**
+   * Initialize the stats file using the sample segment provided.
+   * <br>This involves indexing each row of the sample segment using MutableSegmentImpl. This is equivalent to
+   * consuming the rows of a segment.
+   * Although they will be in a different order than consumed by the host, the stats should be equivalent.
+   * <br>Invoking a destroy on the MutableSegmentImpl at the end will dump the collected stats into the stats.ser
+   * file provided in the statsHistory.
+   */
+  public File initializeStatsHistory() {
+
+    File statsFile = new File(_workingDir, STATS_FILE_NAME);
+    RealtimeSegmentStatsHistory sampleStatsHistory;
+    try {
+      sampleStatsHistory = RealtimeSegmentStatsHistory.deserialzeFrom(statsFile);
+    } catch (IOException | ClassNotFoundException e) {
+      throw new RuntimeException(
+          "Exception when deserializing stats history from stats file " + statsFile.getAbsolutePath(), e);
+    }
+
+    RealtimeIndexOffHeapMemoryManager memoryManager = new DirectMemoryManager(_segmentMetadata.getName());
+    SegmentZKMetadata segmentZKMetadata = getSegmentZKMetadata(_segmentMetadata, _segmentMetadata.getTotalDocs());
+
+    // create a config
+    RealtimeSegmentConfig.Builder realtimeSegmentConfigBuilder =
+        new RealtimeSegmentConfig.Builder().setTableNameWithType(_tableNameWithType)
+            .setSegmentName(_segmentMetadata.getName()).setStreamName(_tableNameWithType)
+            .setSchema(_segmentMetadata.getSchema()).setCapacity(_segmentMetadata.getTotalDocs())
+            .setAvgNumMultiValues(_avgMultiValues).setNoDictionaryColumns(_noDictionaryColumns)
+            .setVarLengthDictionaryColumns(_varLengthDictionaryColumns).setInvertedIndexColumns(_invertedIndexColumns)
+            .setSegmentZKMetadata(segmentZKMetadata).setOffHeap(true).setMemoryManager(memoryManager)
+            .setStatsHistory(sampleStatsHistory);
+
+    // create mutable segment impl
+    MutableSegmentImpl mutableSegmentImpl = new MutableSegmentImpl(realtimeSegmentConfigBuilder.build(), null);
+
+    // read all rows and index them
+    try (PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_sampleCompletedSegment)) {
+      GenericRow row = new GenericRow();
+      while (segmentRecordReader.hasNext()) {
+        row = segmentRecordReader.next(row);
+        mutableSegmentImpl.index(row, null);
+        row.clear();
+      }
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)